### DISCUSSION TITLE
spring kafka test input-output queue

### DISCUSSION THREAD

**Discussion by Vichukano** [spring-projects/spring-kafka#2696] [Ideas] [answered]
**Hallo!**
To write integration tests using Kafka, you often have to write boilerplate code like this:

 ```
var consumerFactory = new DefaultKafkaConsumerFactory<String, String>(props);
var containerProps = new ContainerProperties(TOPIC_OUT);  
var listenerContainer = new KafkaMessageListenerContainer<>(consumerFactory, containerProps);
listenerContainer.setupMessageListener((MessageListener<String, String>) consumerRecords::add);
listenerContainer.start();
ContainerTestUtils.waitForAssignment(listenerContainer, embeddedKafkaBroker.getPartitionsPerTopic());
 ```

It may be possible to come up with some mechanism by which it would be possible to set incoming and outgoing queues in tests, for example like this:

 ```
@OutputQueue(topic = TOPIC_OUT, partitions = 1)
 private BlockingQueue<ConsumerRecord<String, String>> consumerRecords;
/////
ConsumerRecord<String, String> record = consumerRecords.poll(5, TimeUnit.SECONDS);
 ```

Do you think that would be a feature?

**Comment by artembilan** [comment, marked as answer]
My five cents.
I'm not a fan of annotations: they don't give that code flow flexibility when we simply can control everything with respective method calls.
Plus I don't see a correlation between your annotation and whatever we have to provide for some specific container to consume with.

I also wouldn't call this an "integration test" since we create the container manually. 
It would be an integration one, if we verify a `@KafkaListener` actions from a production code.
And to be honest that is exactly what target projects must tests.
We use that `ContainerTestUtils.waitForAssignment()` in the framework unit tests because that is exactly our responsibility to provide coverage for the code we build.
Your project, on the other hand, does not do anything with the framework code: you just do your own logic.
Therefore it is a bit strange to see a framework code verification in the target project.
